### PR TITLE
Add   individual quiz runner

### PR DESCRIPTION
**Pull request by EmilChri**
# Description
This is the Individual quiz runner, related to the retake quiz user story. You can now run a quiz and answer the questions.

## What is added/changed/removed
The quiz runner has been added, the pages are rendered in attempt.php.
When answers are checked it is stored in the session (using JavaScript eventlisteners).
The Pages are made using Mustache Templates.
Both behat tests and unit tests have been made to ensure correct functionality.

## Issue/fixes reference
#124

## Testing
-Both behat and unit tests should run.
-Check that a newly created livequiz can be shown with the demo data (cities in europe quiz).

# Checks
- [x] Updated / added tests for these changes
- [x] PHPunit locally passed
- [x] Codesniffer locally passed
- [x] Behat locally passed

### CODE REVIEW

File: server/moodle/admin/tool/policy/amd/build/jquery-eu-cookie-law-popup.min.js.map
Comment (adomasciplys) [inline server/moodle/admin/tool/policy/amd/build/jquery-eu-cookie-law-popup.min.js.map:1]: For some reason the file was changed. We have copied stagings version of the file into this pull-request. But it still says, for some unknown reason, that we have changed the file 💯

File: server/moodle/mod/livequiz/classes/models/student_quiz_relation.php
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/models/student_quiz_relation.php:1]: Generally against pushing template code, but i'll let it slide this time, due to limited time before codefreeze

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php [hunk 1]
```diff
@@ -83,6 +84,43 @@ public function get_livequiz_instance(int $id): livequiz {
...
+     * Constructs a new question object and appends it to the livequiz object.
+     *
+     * @param livequiz $livequiz
+     * @param string $title
+     * @param string $description
+     * @param int $timelimit
+     * @param string $explanation
+     * @return question
+     */
+    public function new_question(
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/services/livequiz_services.php:97]: What's the point with adding this? PHPStorm claims 'no usages'

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php [hunk 2]
```diff
@@ -83,6 +84,43 @@ public function get_livequiz_instance(int $id): livequiz {
...
+    /**
+     * Constructs a new answer object and appends it to the question object.
+     *
+     * @param question $question
+     * @param int $correct
+     * @param string $description
+     * @param string $explanation
+     * @return answer
+     */
+    public function new_answer(question $question, int $correct, string $description, string $explanation): answer {
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/services/livequiz_services.php:119]: Same here

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php [hunk 3]
```diff
@@ -166,4 +205,19 @@ private function get_questions_with_answers(int $quizid): array {
...
         return $questions;
     }
+    /**
+     * Creates a new participation record in the database.
+     * @param int $studentid
+     * @param int $quizid
+     * @throws dml_exception
+     * @return participation
+     */
+    public function new_participation(int $studentid, int $quizid): participation {
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/services/livequiz_services.php:215]: Same here.
Hvis det er tiltænkt til fremtidig brug af jer, så resolve og ignorér mine kommentarer hvad dette vedrører.
Ellers ser jeg ikke nogen grund til at push'e dem

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php [hunk 4]
```diff
@@ -166,4 +205,19 @@ private function get_questions_with_answers(int $quizid): array {
         }
         return $questions;
     }
+    /**
+     * Creates a new participation record in the database.
+     * @param int $studentid
+     * @param int $quizid
+     * @throws dml_exception
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/services/livequiz_services.php:212]: This function doesnt throw a dml_exception?

File: server/moodle/mod/livequiz/classes/models/question.php [hunk 5]
```diff
@@ -200,4 +201,52 @@ public function add_answer(answer $answer): void {
...
+        $hasmultipleanswers = false;
+        foreach ($this->answers as $answer) {
+            if ($answer->get_correct()) {
+                $numcorrect++;
+            }
+        }
+        if ($numcorrect > 1) {
+            $hasmultipleanswers = true;
+        }
+        return $hasmultipleanswers;
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/classes/models/question.php:221]: For at undgå ønedvendige iterationer. I må selv om i vil bruge det

```suggestion
        // This is a simple check to see if the question has multiple correct answers.
       $numcorrect = 0;

    foreach ($this->answers as $answer) {
        if ($answer->get_correct()) {
            $numcorrect++;
            if ($numcorrect > 1) {
                return true;
            }
        }
    }

    return false;
```

File: server/moodle/mod/livequiz/style.css
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/style.css:1]: What is the point of this?

File: server/moodle/mod/livequiz/version.php [hunk 6]
```diff
@@ -28,6 +28,6 @@
 
 
 $plugin = new stdClass(); // Initialize $plugin as an object.
-$plugin->version = 2024072516;
+$plugin->version = 2024072512;
```
Comment (ChristofferGamel) [inline server/moodle/mod/livequiz/version.php:31]:
```suggestion
$plugin->version = 2024072516;
```

**Review by ChristofferGamel — COMMENTED** [review]
Generelt set et fint PR. Rigtig fedt med gode kommentarer og docblocks.
Men for eftertiden please, om muligt, undgå en PR i den her størrelse. Det er rigtig svært at gennemskue sammenhængen i koden med 4500 linjer tilføjet, og 49 filer ændret.

Jeg har nogle enkelte spørgsmål/foreslag til forbedringer. Når besvaret får i en approval herfra 😃

**Review by mantarias — APPROVED** [review]

**Review by DeeKahy — APPROVED** [review]

**Review by DeeKahy — APPROVED** [review]
Assuming the tests still pass after syncing this pr i still approve.

Please make sure the merge conflicts dont break any of your functionality

**Review by DeeKahy — APPROVED** [review]

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php (hunk 4 above)
Comment (EmilChri) [reply server/moodle/mod/livequiz/classes/services/livequiz_services.php:278]: This is our codesniffer requesting us to have it, do not know why

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php (hunk 2 above)
Comment (EmilChri) [reply server/moodle/mod/livequiz/classes/services/livequiz_services.php:120]: jeg ved ikke hvorfor det står som at det er os der har tilføjet det:
image

Comment (ChristofferGamel) [reply server/moodle/mod/livequiz/classes/services/livequiz_services.php:120]: Hvis i har haft merget noget af vores ind i jeres branch, så er det nok gamle relics, som vi siden har slettet. Vi bruger det ihvertfald ikke til noget

File: server/moodle/mod/livequiz/classes/models/student_quiz_relation.php
Comment (od45pv) [reply server/moodle/mod/livequiz/classes/models/student_quiz_relation.php:1]: Removed.

File: server/moodle/mod/livequiz/classes/services/livequiz_services.php (hunk 4 above)
Comment (od45pv) [reply server/moodle/mod/livequiz/classes/services/livequiz_services.php]: You'll see that in the newest commits there's a transaction being started, that might throw a dml_exception.

**Review by ZeroConfidence — APPROVED** [review]

**Pull request merged**